modules/gt_tree_gen.py

```diff
@@ -1,73 +1,94 @@
 import math
-import random
 import jax.numpy as jnp
+from jax import vmap
+from jax import random
 
 target_depth = 0
 num_mutations = 0
 sequences = []
+n_letters = 0
+
+def mutate(key, exclude_indexes):
+  '''
+    Mutates a sequence by randomly changing one of the letters
+
+    Accepts:
+      exclude_indexes: list of indexes to exclude from the mutation (shape = (n_mutations,1))
+
+    Returns:
+      mutated sequence
+  '''
+  global n_letters
+
+  space = jnp.tile(jnp.arange(n_letters), (num_mutations, 1)) 
+  options  = space[space != exclude_indexes].reshape((space.shape[0],space.shape[1] - exclude_indexes.shape[1]))
+
+  key_ = random.split(key, num_mutations)
+
+  mutations = vmap(random.choice)(key_, options)
+
+  return jnp.expand_dims(mutations, axis=1)
+
+def create_seq(key, seq, depth):
+
+  k1, k2 = random.split(key)
 
-def create_seq(seq, depth):
   print()
   
 
   if(depth > target_depth):
     return
 
-  # index_1 = random.randint(0,len(seq)-1)
-  # index_2 = random.randint(0,len(seq)-1)
-
-  # seq_1 = seq[::]
-  # seq_2 = seq[::]
-  # seq_1[index_1] = int(not(seq[index_1]))
-  # seq_2[index_2] = int(not(seq[index_2]))
-
-
   random_indexes = [[] for x in range(2)]
   
-  random_indexes[0]= random.sample(range(len(seq)), num_mutations)
-  random_indexes[1]= random.sample(range(len(seq)), num_mutations)
+  random_indexes[0]= random.choice(k1, jnp.array(range(len(seq))), (num_mutations,1), replace=False)
+  random_indexes[1]= random.choice(k2, jnp.array(range(len(seq))), (num_mutations,1), replace=False)
+
+
+  seq_1 = seq
+  seq_2 = seq
 
+  seq_1 = seq_1.at[random_indexes[0]].set(mutate(k1, seq[random_indexes[0]]))
+  seq_2 = seq_2.at[random_indexes[1]].set(mutate(k2, seq[random_indexes[1]]))
 
-  seq_1 = seq[::]
-  seq_2 = seq[::]
-  for k in range(0,num_mutations):
-    seq_1[random_indexes[0][k]] = int(not(seq[random_indexes[0][k]]))
-    seq_2[random_indexes[1][k]] = int(not(seq[random_indexes[1][k]]))
 
   #seq
   print((depth+1)*'|--->',end="")
-  print(seq_1)
+  print(list(seq_1.tolist()))
   sequences[depth].append(seq_1)
-  create_seq(seq_1, depth+1)
+  create_seq(k1, seq_1, depth+1)
   
 
   #new_seq
   print((depth+1)*'|--->',end="")
-  print(seq_2)
+  print(list(seq_2.tolist()))
   sequences[depth].append(seq_2)
-  create_seq(seq_2, depth+1)
+  create_seq(k2, seq_2, depth+1)
 
 def generate_groundtruth(metadata, seed = 42):
-    random.seed(seed)
+    key = random.PRNGKey(seed)
```

Comment:
/wtd does this preserve the same random number generation as with python's standard library 🤔